src/referenceHandling.ts

```diff
@@ -0,0 +1,172 @@
+// (C) 2007-2019 GoodData Corporation
+import { VisualizationObject } from '@gooddata/typings';
+import { cloneDeep, get, pickBy, set } from 'lodash';
+import { v4 as uuid } from 'uuid';
+
+import {
+    IReferenceProperties,
+    IColorMappingReferenceProperty,
+    IColorMappingUriProperty,
+    IUriProperties
+} from './interfaces';
+
+type IVisualizationObject = VisualizationObject.IVisualizationObject;
+type IReferenceItems = VisualizationObject.IReferenceItems;
+
+/**
+ * Constants
+ */
+const CONTROLS_PROPERTY_PATH = ['controls'];
+const COLOR_MAPPING_PROPERTY_PATH = [...CONTROLS_PROPERTY_PATH, 'colorMapping'];
+
+/**
+ * Helpers
+ */
+const getReferenceValue = (id: string, references: IReferenceItems) => references[id];
+const getReferenceId = (value: string, references: IReferenceItems) =>
+    Object.keys(references).find(id => references[id] === value);
+
+const defaultIdGenerator = () => uuid().replace(/-/g, '');
+
+const createConverter = <TInProperties, TOutProperties>(
+    conversionBody:
+        (properties: TInProperties, references: IReferenceItems, idGenerator: () => string) => {
+            convertedProperties: TOutProperties, convertedReferences: IReferenceItems
+        }
+) => (mdObject: IVisualizationObject, idGenerator: () => string = defaultIdGenerator) => {
+    const { content } = mdObject;
+    if (!content) {
+        return mdObject;
+    }
+
+    const { properties } = content;
+    if (!properties) {
+        return mdObject;
+    }
+
+    // prepare result objects
+    const convertedMdObject: IVisualizationObject = cloneDeep(mdObject);
```

Comment:
This is a textbook use-case for immer.

Reply:
`immer` is not used in this project (yet). IDK if adding it would be justified 😟

Reply:
I wrote around the need for `immer` in the end.